cmd/kubeadm/app/images/puller.go

```diff
@@ -0,0 +1,57 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package images
+
+import (
+	"fmt"
+
+	kubeadmdefaults "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha1"
+	utilsexec "k8s.io/utils/exec"
+)
+
+// Puller is an interface for pulling images
+type Puller interface {
+	Pull(string) error
+}
+
+// ImagePuller is a struct that can pull images and hides the implementation (crictl vs docker)
+type ImagePuller struct {
+	criSocket  string
+	exec       utilsexec.Interface
+	crictlPath string
+}
+
+// NewImagePuller returns a ready to go ImagePuller
+func NewImagePuller(execer utilsexec.Interface, criSocket string) (*ImagePuller, error) {
+	crictlPath, err := execer.LookPath("crictl")
+	if crictlPath == "" && err != nil {
```

Comment:
What happens if `crictlPath` is "" but there isn't an error?

Reply:
`&&` -> `||`

Reply:
Also wrong logic from before -- If we *need* crictl (our crisocket is not the default) then we fail. Otherwise it's fine that we don't have crictl available (since we use docker instead).